core/types/subtyping.cc

```diff
@@ -221,7 +258,9 @@ TypePtr Types::lub(const GlobalState &gs, const TypePtr &t1, const TypePtr &t2)
         newTargs.reserve(indexes.size());
         // code below inverts permutation of type params
         int j = 0;
-        bool changed = false;
+        bool changedFromT2 = false;
+        // If klasses are equal, then it's possible that t1s <: t2s.
+        bool changedFromT1 = a1->klass != a2->klass;
```

Comment:
This is the correctness fix.

Before this change:
* `lub(T::Array[T.untyped], T::Array[X])` would return a `T::Array[T.untyped]` that is pointer equivalent with the first parameter.
* `lub(T::Array[X], T::Array[T.untyped])` would return a fresh `T::Array[T.untyped]` that is not pointer equivalent with the second parameter.

After this change:
* Both `lub(T::Array[T.untyped], T::Array[X])` and `lub(T::Array[X], T::Array[T.untyped])` return the same `T::Array[T.untyped]` pointer passed in to lub.